File: python/85_Maximal_Rectangle.py
class Solution:
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        n = len(matrix)-1 #行
        if n>=0:
            m = len(matrix[0])-1 #列
        else:
            return 0
         
        tb = set()
        i = 0
        while i<=n:
            j = 0
            while j<=m:
                if matrix[i][j] == "1":
                    tb.add((i,j))
                j += 1
            i += 1

        #print(tb)
        ans = 1 if len(tb) > 0 else 0
        for element in tb:
            i = element[0]
            j = element[1]
            #print("====", i , j)

            x = 1
            while i+1 <= n:
                if matrix[i+1][j] == "0":
                    break
                else:
                    x += 1
                    i += 1

            y = 1
            i = element[0]
            while j+1 <= m:
                if matrix[i][j+1] == "0":
                    break
                else:
                    y += 1
                    j += 1

            #print(element, x, y)     
            ans = max(ans, x*y)

        return ans




sol = Solution()
matrix = [
  ["1","0","1","0","0"],
  ["1","0","1","1","1"],
  ["1","1","1","1","1"],
  ["1","0","0","1","0"]
]
print(sol.maximalRectangle(matrix))
                

File: cpp/85_maximalRectangle.cpp
#include <string>  
#include <vector>
#include <iostream>  
#include <unordered_map>  
#include <unordered_set> 
#include <list>
#include <algorithm>
#include <queue>
#include <stack>
#include <limits.h>
using namespace std;

class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if(matrix.size()==0) return 0;
        int ans=0;
        vector<int> heights(matrix[0].size());
        for(int i=0;i<matrix.size();++i){
            for(int j=0;j<matrix[0].size();++j){
                if(matrix[i][j]=='1') heights[j]++;
                else heights[j]=0;
            }
            ans = max(ans,largestRectangleArea(heights));
        }

        return ans;
    }

    int largestRectangleArea(vector<int>& heights) {
        if(heights.size()==0) return 0;
        int ans=0;
        stack<int> s;
        heights.push_back(0);
        for(int i=0;i<heights.size();++i){
            while(!s.empty() && heights[s.top()]>=heights[i]){
                int cur = s.top();
                s.pop();
                ans = max(ans, heights[cur]*(s.empty()?i:(i-s.top()-1)));
            }

            s.push(i);
        }

        return ans;
    }
};

int main(int argc, char const *argv[])
{
    vector<vector<char>> matrix=
                    {{'0','1','1','0','1'},
                        {'1','1','0','1','0'},
                        {'0','1','1','1','0'},
                        {'1','1','1','1','0'},
                        {'1','1','1','1','1'},
                        {'0','0','0','0','0'}};

    Solution sol;
    cout << sol.maximalRectangle(matrix) << endl;
    return 0;
}

